ubi_manifest/app/utils.py

```diff
@@ -0,0 +1,112 @@
+    app_conf: Any, provided_repos: list[str]
+) -> list[dict[str, Any]]:
+    """
+    Returns a list of all {"repo_group": ["repo1, repo2], "url" = "https://config"}
+    items which are then used for creation of depsolving tasks.
+    """
+    all_items: list[dict[str, Any]] = []
+
+    # client-tools repos and ubi repos need to be handled separately
+    client_tools_repos = [repo for repo in provided_repos if "client-tools" in repo]
+    ubi_repos = [repo for repo in provided_repos if "ubi" in repo]
+
+    if client_tools_repos:
+        all_items.extend(
+            get_client_tools_items(
+                client_tools_repos, app_conf.content_config["client-tools"]
+            )
+        )
+    if ubi_repos:
+        all_items.extend(get_ubi_items(ubi_repos, app_conf))
+
+    _LOG.info("Determined items for depsolving: %s", all_items)
+    return all_items
+
+
+def get_client_tools_items(
+    provided_repo_ids: list[str], config_url: str
+) -> list[dict[str, Any]]:
+    """
+    Returns a list of items for depsolving of client-tools repos.
+    For these we don't use full depsolving so it is performed only
+    on the provided repo and no group is needed to be determined.
+    """
+    items: list[dict[str, Any]] = []
+    for repo in provided_repo_ids:
+        items.append({"repo_group": [repo], "url": config_url})
+    return items
+
+
+def get_ubi_items(provided_repo_ids: list[str], app_conf: Any) -> list[dict[str, Any]]:
+    """
+    Returns a list of items for depsolving of ubi repos.
+    """
+    items: list[dict[str, Any]] = []
+    needed_repo_groups: dict[str, set[str]] = {}
+
+    config_url = app_conf.content_config["ubi"]
+    allowed_ubi_repo_groups = get_allowed_ubi_repo_groups(config_url, app_conf)
+
+    # Compare provided repo_ids with the generated allowed repo groups and pick
+    # those needed for this request
+    for repo_id in provided_repo_ids:
+        for group_key, repo_group in allowed_ubi_repo_groups.items():
+            if repo_id in repo_group:
+                needed_repo_groups.setdefault(group_key, repo_group)
+                break
+
+    # Create items for depsolving
+    for repo_group in needed_repo_groups.values():
+        items.append({"repo_group": list(sorted(repo_group)), "url": config_url})
+
+    return items
+
+
+def get_allowed_ubi_repo_groups(config_url: str, app_conf: Any) -> dict[str, set[str]]:
+    """
+    Determines the allowed repo groups based on the the available ubi config and Pulp data.
+    """
+    allowed_ubi_repo_groups: dict[str, set[str]] = {}
+
+    _LOG.info("Loading config from %s", config_url)
+
+    loader = ubiconfig.get_loader(config_url)
+    configs = loader.load_all()
+    # Use only configs for major versions
+    configs = [conf for conf in configs if "." not in conf.version]
+
+    with make_pulp_client(app_conf) as client:
+        for conf in configs:
+            repos = get_repo_ids_from_cs(client, conf.content_sets.rpm.output)
```

Comment:
This theoretically can pull more repositories than we want.
If, hypothetically, we would like to populate DOT repos for UBI, we would need to filter this output more precisely in order to get correct repo groups (DOT and non-DOT repo groups would need to be depsolved separately). But I think for now it's good enough as there is no known requirement for it.